src/core/ddsc/src/dds_write.c

```diff
@@ -382,233 +396,192 @@ static bool fill_iox_chunk(dds_writer *wr, const void *sample, void *iox_chunk,
-        // TODO: proper fallback logic to network path?
-
-        // iceoryx_available = false;
-        ret = DDS_RETCODE_OUT_OF_RESOURCES;
-        goto finalize_write;      
-      } 
-    } else {
-      // The user already provided an iceoryx_chunk with data (by using the dds_loan API)
-      // We assume if we got the data from a loan it contains raw data
-      // (i.e. not serialized)
-      // This requires the user to adhere to the contract, we cannot enforce this
-      // with the given API.
-      iox_chunk = (void*) data;
-      iceoryx_header_t *iox_header = iceoryx_header_from_chunk(iox_chunk);
-      iox_header->shm_data_state = IOX_CHUNK_CONTAINS_RAW_DATA;      
-    }
-  }
+  //note: whether the data was loaned cannot be determined in the non-iceoryx case currently
+  if ((ret = get_iox_chunk (wr, data, &iox_chunk)) != 0)
+    return ret;
+  assert (iox_chunk != NULL);
 
   // The following holds from here
   // 1) data still points to the original data
-  // 2) iox_chunk is NULL or
-  // 3) iox_chunk is NOT NULL
+  // 2) iox_chunk is NOT NULL
   //    a) was created by the write call or
   //    b) data is an iceoryx chunk and iox_chunk == data
-  // in case 3 a) iox_chunk will contain serialized or raw data
+  // in case 2 a) iox_chunk will contain serialized or raw data
   //
-  // in case 3 a) and b) we must ensure we publish or release the chunk (failure)
-  // in 3b) we could argue to not release the chunk but the user effectively passed ownership
+  // in case 2 a) and b) we must ensure we publish or release the chunk (failure)
+  // in 2b) we could argue to not release the chunk but the user effectively passed ownership
   //        by calling write
 
   // ddsi_wr->as can be changed by the matching/unmatching of proxy readers if we don't hold the lock
   // it is rather unfortunate that this then means we have to lock here to check, then lock again to
   // actually distribute the data, so some further refactoring is needed.
   ddsrt_mutex_lock (&ddsi_wr->e.lock);
-  bool no_network_readers = addrset_empty (ddsi_wr->as);
+  const bool no_network_readers = addrset_empty (ddsi_wr->as);
   ddsrt_mutex_unlock (&ddsi_wr->e.lock);
-  bool use_only_iceoryx =
-      iceoryx_available && no_network_readers &&
-      ddsi_wr->xqos->durability.kind == DDS_DURABILITY_VOLATILE;
+
+  // NB: local readers are not in L := ddsi_wr->rdary if they use iceoryx.
+  // Furthermore, all readers that ignore local publishers will not use iceoryx.
+  // We will never use only(!) iceoryx if there is any local reader in L.
+  // We will serialize the data in this case and deliver it mixed, i.e.
+  // partially with iceoryx as required by the QoS and type. The readers in L
+  // will get the data via the local delivery mechanism (fast path or slow
+  // path).
+  const bool use_only_iceoryx =
+      no_network_readers &&
+      ddsi_wr->xqos->durability.kind == DDS_DURABILITY_VOLATILE &&
+      ddsi_wr->rdary.n_readers == 0;
```

Comment:
I am a little bit worried about this access outside all locks, but at the same time I suspect it will actually work fine in practice. If unlocked access is indeed fine, it should become an `ddsrt_atomic_uint32_t`, but if you then look at the definition, loading it really is nothing special on any machine.

Perhaps I should sleep on this detail for a night ...

Reply:
I somewhat agree. I usually use another style of programming where the locks are tightly coupled to the structures they are supposed to protect (otherwise it gets unmaintainable fast). This is easier in languages that allow encapsulation though. Then it can be made impossible to use wrongly (without lock).

In this case we should think about a dedicated lock for this whole fast-path structure, unless we can argue it is only modified/read under writer lock anyway. This `n_readers` should only be accessed under lock then, by some `get` function. I can change it like this.

Reply:
If you time to do that, that'd be great!

By the way, there is a lock: https://github.com/eclipse-cyclonedds/cyclonedds/blob/3e5a4ab093cf18e9e7a6887d6b0a00c82b7b1240/src/core/ddsi/include/dds/ddsi/q_entity.h#L233